llvm/lib/CAS/ActionCaches.cpp

```diff
@@ -0,0 +1,225 @@
+#include "llvm/CAS/CASDB.h"
+#include "llvm/CAS/HashMappedTrie.h"
+#include "llvm/CAS/OnDiskHashMappedTrie.h"
+#include "llvm/Support/Alignment.h"
+#include "llvm/Support/BLAKE3.h"
+#include "llvm/Support/Path.h"
+#include <memory>
+
+#define DEBUG_TYPE "action-caches"
+
+using namespace llvm;
+using namespace llvm::cas;
+
+namespace {
+
+using HasherT = BLAKE3;
+using HashType = decltype(HasherT::hash(std::declval<ArrayRef<uint8_t> &>()));
+
+template <size_t Size> class CacheEntry {
+public:
+  CacheEntry() = default;
+  CacheEntry(ArrayRef<uint8_t> Hash) { llvm::copy(Hash, Value.data()); }
+  CacheEntry(const CacheEntry &Entry) { llvm::copy(Entry.Value, Value.data()); }
+  ArrayRef<uint8_t> getValue() const { return Value; }
+
+private:
+  std::array<uint8_t, Size> Value;
+};
+
+class InMemoryActionCache final : public ActionCache {
+public:
+  InMemoryActionCache(CASDB &CAS);
+
+  Error putImpl(ArrayRef<uint8_t> ActionKey, const ObjectRef &Result) final;
+  Optional<ObjectRef> getImpl(ArrayRef<uint8_t> ActionKey) const final;
+  Error poisonImpl(ArrayRef<uint8_t> Key) final;
+
+private:
+  using DataT = CacheEntry<sizeof(HashType)>;
+  using InMemoryCacheT = ThreadSafeHashMappedTrie<DataT, sizeof(HashType)>;
+
+  InMemoryCacheT Cache;
+};
+
+class OnDiskActionCache final : public ActionCache {
+public:
+  Error putImpl(ArrayRef<uint8_t> ActionKey, const ObjectRef &Result) final;
+  Optional<ObjectRef> getImpl(ArrayRef<uint8_t> ActionKey) const final;
+  Error poisonImpl(ArrayRef<uint8_t> Key) final;
+
+  static Expected<std::unique_ptr<OnDiskActionCache>> create(CASDB &CAS,
+                                                             StringRef Path);
+
+private:
+  static StringRef getHashName() { return "BLAKE3"; }
+  static StringRef getActionCacheTableName() {
+    static const std::string Name =
+        ("llvm.actioncache[" + getHashName() + "->" + getHashName() + "]")
+            .str();
+    return Name;
+  }
+  static StringRef getDataPoolTableName() {
+    static const std::string Name =
+        ("llvm.actioncache.data[" + getHashName() + "]").str();
+    return Name;
+  }
+  static constexpr StringLiteral DataPoolFile = "data";
+  static constexpr StringLiteral ActionCacheFile = "actions";
+  static constexpr StringLiteral FilePrefix = "v1.";
+
+  OnDiskActionCache(CASDB &CAS, StringRef RootPath,
+                    OnDiskDataAllocator DataPool,
+                    OnDiskHashMappedTrie ActionCache);
+
+  std::string Path;
+  OnDiskDataAllocator DataPool;
+  OnDiskHashMappedTrie Cache;
+  using DataT = CacheEntry<sizeof(HashType)>;
+  using ActionCacheResultT = std::atomic<DataT *>;
+};
+} // end namespace
+
+// TODO: Check the hash schema is the same between action cache and CAS. If we
+// can derive that from static type information, that would be even better.
+InMemoryActionCache::InMemoryActionCache(CASDB &CAS) : ActionCache(CAS) {}
+
+OnDiskActionCache::OnDiskActionCache(CASDB &CAS, StringRef Path,
```

Comment:
Style nit: maybe group `OnDiskActionCache` and `InMemoryActionCache` methods rather than interleaving them?